# Week 1 Assignment: Housing Prices

In this exercise you'll try to build a neural network that predicts the price of a house according to a simple formula.

Imagine that house pricing is as easy as:

A house has a base cost of 50k, and every additional bedroom adds a cost of 50k. This will make a 1 bedroom house cost 100k, a 2 bedroom house cost 150k etc.

How would you create a neural network that learns this relationship so that it would predict a 7 bedroom house as costing close to 400k etc.

Hint: Your network might work better if you scale the house price down. You don't have to give the answer 400...it might be better to create something that predicts the number 4, and then your answer is in the 'hundreds of thousands' etc.

In [22]:
import tensorflow as tf
from keras.layers import Dense
from keras.models import Sequential
import numpy as np

In [23]:
number_of_bedrooms = [1,2,3,4,5,6]
#up to 6 bedrooms

#start an empty list
house_cost = []

#fill the list with the house costs
for i in number_of_bedrooms:
    house_cost.append(50 + i*50)
    #costs in thousands of dollars

#convert the lists to NumPy arrays:
house_cost_array = np.array(house_cost, dtype = float)
number_of_bedrooms_array = np.array(number_of_bedrooms, dtype = float)

print(number_of_bedrooms_array)
print(house_cost_array)

[1. 2. 3. 4. 5. 6.]
[100. 150. 200. 250. 300. 350.]


In [25]:
# GRADED FUNCTION: house_model
def house_model(number_of_bedrooms, house_cost, EPOCHS):
    ### START CODE HERE
    
    # Define input and output tensors with the values for houses with 1 up to 6 bedrooms
    # Hint: Remember to explictly set the dtype as float
    xs = number_of_bedrooms
    ys = house_cost
    
    # Define your model (should be a model with 1 dense layer and 1 unit)
    model = Sequential()
    model.add(Dense(units=1, input_shape=[1]))
    
    # Compile your model
    # Set the optimizer to Stochastic Gradient Descent
    # and use Mean Squared Error as the loss function
    model.compile(optimizer='sgd', loss='mean_squared_error')
    
    # Train your model for 1000 epochs by feeding the i/o tensors
    model.fit(xs, ys, epochs = EPOCHS)
    
    ### END CODE HERE
    return model

## As a first approach, let's use the sklearn.StandardScaler() to normalize the data

In [24]:
from sklearn.preprocessing import StandardScaler

#scale the house costs
scaler = StandardScaler()

scaler.fit(house_cost_array.reshape(-1,1))
house_cost_scaled = scaler.transform(house_cost_array.reshape(-1,1))

print(house_cost_scaled)

[[-1.46385011]
 [-0.87831007]
 [-0.29277002]
 [ 0.29277002]
 [ 0.87831007]
 [ 1.46385011]]


Now that you have a function that returns a compiled and trained model when invoked, use it to get the model to predict the price of houses: 

In [26]:
# Get your trained model
model = house_model(number_of_bedrooms = number_of_bedrooms_array, house_cost = house_cost_scaled, EPOCHS = 1000)

Epoch 1/1000
1/1 [==============================] - 0s 172ms/step - loss: 34.4623
Epoch 2/1000
1/1 [==============================] - 0s 2ms/step - loss: 16.2512
Epoch 3/1000
1/1 [==============================] - 0s 2ms/step - loss: 7.8204
Epoch 4/1000
1/1 [==============================] - 0s 2ms/step - loss: 3.9161
Epoch 5/1000
1/1 [==============================] - 0s 1ms/step - loss: 2.1070
Epoch 6/1000
1/1 [==============================] - 0s 1ms/step - loss: 1.2675
Epoch 7/1000
1/1 [==============================] - 0s 2ms/step - loss: 0.8768
Epoch 8/1000
1/1 [==============================] - 0s 2ms/step - loss: 0.6939
Epoch 9/1000
1/1 [==============================] - 0s 1ms/step - loss: 0.6072
Epoch 10/1000
1/1 [==============================] - 0s 1ms/step - loss: 0.5649
Epoch 11/1000
1/1 [==============================] - 0s 1ms/step - loss: 0.5433
Epoch 12/1000
1/1 [==============================] - 0s 2ms/step - loss: 0.5312
Epoch 13/1000
1/1 [==========================

Now that your model has finished training it is time to test it out! You can do so by running the next cell.

In [27]:
new_y = 7.0
prediction = model.predict([new_y])[0]
print(prediction)

[2.0212028]


In [28]:
#Scale back data to its original representation

prediction_re_scaled = scaler.inverse_transform(np.array(prediction).reshape(-1,1))
print(prediction_re_scaled)

[[397.59305]]


When we re-scale the data using the inverse_transform of scaler function, we get 397.6 k, which is pretty close to 400k.

## Let's compare the results of the function when we simply normalize the Y values dividing them by 10

In [31]:
house_cost_normalized = house_cost_array/100
print(house_cost_normalized)

[1.  1.5 2.  2.5 3.  3.5]


In [35]:
#reshape the Ys to fit the neural network
house_cost_normalized = house_cost_normalized.reshape(-1,1)
print(house_cost_normalized)

[[1. ]
 [1.5]
 [2. ]
 [2.5]
 [3. ]
 [3.5]]


In [32]:
model = house_model(number_of_bedrooms = number_of_bedrooms_array, house_cost = house_cost_normalized, EPOCHS = 1000)

Epoch 1/1000
1/1 [==============================] - 0s 162ms/step - loss: 72.5317
Epoch 2/1000
1/1 [==============================] - 0s 2ms/step - loss: 33.5699
Epoch 3/1000
1/1 [==============================] - 0s 1ms/step - loss: 15.5372
Epoch 4/1000
1/1 [==============================] - 0s 1ms/step - loss: 7.1911
Epoch 5/1000
1/1 [==============================] - 0s 1ms/step - loss: 3.3283
Epoch 6/1000
1/1 [==============================] - 0s 2ms/step - loss: 1.5404
Epoch 7/1000
1/1 [==============================] - 0s 1ms/step - loss: 0.7130
Epoch 8/1000
1/1 [==============================] - 0s 1ms/step - loss: 0.3300
Epoch 9/1000
1/1 [==============================] - 0s 1ms/step - loss: 0.1528
Epoch 10/1000
1/1 [==============================] - 0s 1ms/step - loss: 0.0707
Epoch 11/1000
1/1 [==============================] - 0s 1ms/step - loss: 0.0328
Epoch 12/1000
1/1 [==============================] - 0s 1ms/step - loss: 0.0152
Epoch 13/1000
1/1 [=========================

In [33]:
new_y = 7.0
prediction = model.predict([new_y])[0]
print(prediction)

[4.0002537]


### **The value obtained is approximately equals to 4**

If everything went as expected you should see a prediction value very close to 4. **If not, try adjusting your code before submitting the assignment.** Notice that you can play around with the value of `new_y` to get different predictions. In general you should see that the network was able to learn the linear relationship between `x` and `y`, so if you use a value of 8.0 you should get a prediction close to 4.5 and so on.

**Congratulations on finishing this week's assignment!**

You have successfully coded a neural network that learned the linear relationship between two variables. Nice job!

**Keep it up!**